File: api/models/report_spec.py
"""The shape of a one-page executive summary, as the model fills it and the browser draws it.

A REPORT IS A SPEC, NOT A DOCUMENT. The model gathers figures with execute_sql and
then fills THIS object; a deterministic renderer in the browser
(utils/reportPdf.ts) lays it out on exactly one page with the shop's header. The
guardrails the owner asked for -- one page, minimal prose, the AI-inferred title
top-right -- are therefore properties of a schema and a renderer, not requests
made of a model.

STRICT-SCHEMA SHAPE, ON PURPOSE. Ollama's `format` and Anthropic's structured
output both want every object closed (additionalProperties: false) with every
property required, which rules out free-form dicts: a table row is a LIST of
cells aligned with `columns`, a chart is a list of {label, value} points that the
handler turns into a chart_config, and every optional field is nullable rather
than absent. postprocess.strictify also refuses `$defs`, so `model_json_schema`
is overridden to inline the nested models -- keep them flat enough that this
stays a mechanical step.

The caps are the one-page budget expressed as data: four KPI tiles, four blocks,
eight rows by five columns, twelve points, one text block of two sentences. The
renderer still measures and drops what does not fit, naming what it dropped;
these numbers make that rare rather than impossible.

The mirror is utils/reportSpec.ts; __tests__/fixtures/reportSpecExample.json is
validated by both sides so they cannot drift apart unnoticed.
"""
from __future__ import annotations

import copy
from datetime import date
from typing import Any, Literal

from pydantic import BaseModel, Field, model_validator

Format = Literal["currency", "integer", "percent", "plain"]
Cell = str | float | None

TITLE_MAX = 40
HEADLINE_MAX = 200
KPI_MAX = 4
BLOCK_MAX = 4
TABLE_ROWS_MAX = 8
TABLE_COLUMNS_MAX = 5
CHART_POINTS_MIN = 3
CHART_POINTS_MAX = 12
TEXT_BODY_MAX = 240
NOTE_MAX = 120


class Kpi(BaseModel):
    label: str = Field(min_length=1, max_length=24)
    value: float
    format: Format
    caption: str | None = Field(default=None, max_length=32)


class Column(BaseModel):
    label: str = Field(min_length=1, max_length=20)
    format: Format


class TableBlock(BaseModel):
    type: Literal["table"]
    title: str = Field(min_length=1, max_length=TITLE_MAX)
    columns: list[Column] = Field(min_length=1, max_length=TABLE_COLUMNS_MAX)
    rows: list[list[Cell]] = Field(min_length=1, max_length=TABLE_ROWS_MAX)
    total_row: list[Cell] | None = None
    note: str | None = Field(default=None, max_length=NOTE_MAX)

    @model_validator(mode="after")
    def _rows_match_columns(self) -> "TableBlock":
        width = len(self.columns)
        for row in self.rows:
            if len(row) != width:
                raise ValueError(f"every row must have {width} cells, one per column")
        if self.total_row is not None and len(self.total_row) != width:
            raise ValueError(f"total_row must have {width} cells, one per column")
        return self


class ChartPoint(BaseModel):
    label: str = Field(min_length=1, max_length=40)
    value: float


class ChartBlock(BaseModel):
    type: Literal["chart"]
    title: str = Field(min_length=1, max_length=TITLE_MAX)
    chart_type: Literal["area", "bar", "bar_horizontal", "pie"]
    x_label: str = Field(min_length=1, max_length=24)
    y_label: str = Field(min_length=1, max_length=24)
    points: list[ChartPoint] = Field(min_length=CHART_POINTS_MIN, max_length=CHART_POINTS_MAX)
    note: str | None = Field(default=None, max_length=NOTE_MAX)


class TextBlock(BaseModel):
    type: Literal["text"]
    body: str = Field(min_length=1, max_length=TEXT_BODY_MAX)


Block = TableBlock | ChartBlock | TextBlock


def inline_refs(schema: dict[str, Any]) -> dict[str, Any]:
    """Resolve every `$ref` into a copy of its definition and drop `$defs`.

    strictify() refuses `$defs` because strict-mode `$ref` support is uneven
    across vendors. Pydantic emits one per nested model, so the report's schema
    is inlined here, once, rather than each nested type being hand-flattened.
    """
    defs = schema.get("$defs") or {}

    def resolve(node: Any) -> Any:
        if isinstance(node, list):
            return [resolve(n) for n in node]
        if not isinstance(node, dict):
            return node
        ref = node.get("$ref")
        if isinstance(ref, str) and ref.startswith("#/$defs/"):
            target = copy.deepcopy(defs[ref[len("#/$defs/"):]])
            # Sibling keys beside a $ref (a description, a title) ride along.
            merged = {**resolve(target), **{k: v for k, v in node.items() if k != "$ref"}}
            return merged
        return {k: resolve(v) for k, v in node.items() if k != "$defs"}

    return resolve(schema)


class ReportSpec(BaseModel):
    title: str = Field(min_length=1, max_length=TITLE_MAX)
    period_start: date
    period_end: date
    period_label: str = Field(min_length=1, max_length=40)
    headline: str = Field(min_length=1, max_length=HEADLINE_MAX)
    kpis: list[Kpi] = Field(max_length=KPI_MAX)
    blocks: list[Block] = Field(min_length=1, max_length=BLOCK_MAX)

    @model_validator(mode="after")
    def _one_page_of_prose(self) -> "ReportSpec":
        if sum(1 for b in self.blocks if isinstance(b, TextBlock)) > 1:
            raise ValueError("at most one text block: numbers over words")
        if self.period_start > self.period_end:
            raise ValueError("period_start must not be after period_end")
        return self

    @classmethod
    def model_json_schema(cls, *args: Any, **kwargs: Any) -> dict[str, Any]:  # type: ignore[override]
        return inline_refs(super().model_json_schema(*args, **kwargs))


__all__ = [
    "BLOCK_MAX", "CHART_POINTS_MAX", "CHART_POINTS_MIN", "HEADLINE_MAX", "KPI_MAX",
    "NOTE_MAX", "TABLE_COLUMNS_MAX", "TABLE_ROWS_MAX", "TEXT_BODY_MAX", "TITLE_MAX",
    "Block", "Cell", "ChartBlock", "ChartPoint", "Column", "Format", "Kpi", "ReportSpec",
    "TableBlock", "TextBlock", "inline_refs",
]

File: utils/reportSpec.ts
import type { ChartConfig } from '@/utils/insightsAccess';
import type { ValueFormat } from '@/utils/chartFormat';

/**
 * The browser's mirror of api/models/report_spec.py.
 *
 * Two shapes reach this file and both are accepted: the spec as the model filled
 * it (a chart block carries `points`) and the result as the handler stored it (a
 * chart block carries a ready `chart_config`, the points having been through the
 * chat gate). `reportSpecOf` normalises both to `chart_config`, so the renderer
 * has one shape to draw and __tests__/fixtures/reportSpecExample.json -- the file
 * pytest validates against the Python model -- also validates here.
 */

export interface ReportKpi {
  label: string;
  value: number;
  format: ValueFormat;
  caption: string | null;
}

export interface ReportColumn {
  label: string;
  format: ValueFormat;
}

export type ReportCell = string | number | null;

export interface ReportTableBlock {
  type: 'table';
  title: string;
  columns: ReportColumn[];
  rows: ReportCell[][];
  total_row: ReportCell[] | null;
  note: string | null;
}

export interface ReportChartBlock {
  type: 'chart';
  title: string;
  chart_config: ChartConfig;
  note: string | null;
}

export interface ReportTextBlock {
  type: 'text';
  body: string;
}

export type ReportBlock = ReportTableBlock | ReportChartBlock | ReportTextBlock;

export interface ReportSpec {
  title: string;
  period_start: string;
  period_end: string;
  period_label: string;
  headline: string;
  kpis: ReportKpi[];
  blocks: ReportBlock[];
}

const FORMATS: ReadonlySet<string> = new Set(['currency', 'integer', 'percent', 'plain']);
const CHART_TYPES: ReadonlySet<string> = new Set(['area', 'pie', 'bar', 'bar_horizontal', 'sparkline']);

const isRecord = (v: unknown): v is Record<string, unknown> =>
  !!v && typeof v === 'object' && !Array.isArray(v);
const isFormat = (v: unknown): v is ValueFormat => typeof v === 'string' && FORMATS.has(v);
const isCell = (v: unknown): v is ReportCell =>
  v === null || typeof v === 'string' || (typeof v === 'number' && Number.isFinite(v));
const optionalText = (v: unknown): string | null | undefined =>
  v === null || v === undefined ? null : typeof v === 'string' ? v : undefined;

function kpiOf(raw: unknown): ReportKpi | null {
  if (!isRecord(raw)) return null;
  const caption = optionalText(raw.caption);
  if (typeof raw.label !== 'string' || typeof raw.value !== 'number' || !isFormat(raw.format)) return null;
  if (caption === undefined) return null;
  return { label: raw.label, value: raw.value, format: raw.format, caption };
}

function chartConfigOf(raw: unknown): ChartConfig | null {
  if (!isRecord(raw)) return null;
  if (typeof raw.chart_type !== 'string' || !CHART_TYPES.has(raw.chart_type)) return null;
  if (!Array.isArray(raw.data) || typeof raw.x_key !== 'string' || typeof raw.y_key !== 'string') return null;
  return raw as unknown as ChartConfig;
}

/** A model-shaped chart block (`points`) as the renderer's chart_config. */
function chartConfigFromPoints(raw: Record<string, unknown>): ChartConfig | null {
  if (!Array.isArray(raw.points) || typeof raw.chart_type !== 'string' || !CHART_TYPES.has(raw.chart_type)) {
    return null;
  }
  const data: Record<string, unknown>[] = [];
  for (const p of raw.points) {
    if (!isRecord(p) || typeof p.label !== 'string' || typeof p.value !== 'number') return null;
    data.push({ label: p.label, value: p.value });
  }
  return {
    chart_type: raw.chart_type as ChartConfig['chart_type'],
    data,
    x_key: 'label',
    y_key: 'value',
    x_label: typeof raw.x_label === 'string' ? raw.x_label : undefined,
    y_label: typeof raw.y_label === 'string' ? raw.y_label : undefined,
  };
}

function blockOf(raw: unknown): ReportBlock | null {
  if (!isRecord(raw) || typeof raw.title !== 'string' && raw.type !== 'text') return null;
  const note = optionalText(raw.note);
  if (raw.type === 'text') {
    return typeof raw.body === 'string' ? { type: 'text', body: raw.body } : null;
  }
  if (raw.type === 'chart') {
    if (note === undefined) return null;
    const chart_config = chartConfigOf(raw.chart_config) ?? chartConfigFromPoints(raw);
    return chart_config ? { type: 'chart', title: raw.title as string, chart_config, note } : null;
  }
  if (raw.type === 'table') {
    if (note === undefined || !Array.isArray(raw.columns) || !Array.isArray(raw.rows)) return null;
    const columns: ReportColumn[] = [];
    for (const c of raw.columns) {
      if (!isRecord(c) || typeof c.label !== 'string' || !isFormat(c.format)) return null;
      columns.push({ label: c.label, format: c.format });
    }
    const rows: ReportCell[][] = [];
    for (const r of raw.rows) {
      if (!Array.isArray(r) || r.length !== columns.length || !r.every(isCell)) return null;
      rows.push(r as ReportCell[]);
    }
    let total_row: ReportCell[] | null = null;
    if (raw.total_row !== null && raw.total_row !== undefined) {
      if (!Array.isArray(raw.total_row) || raw.total_row.length !== columns.length || !raw.total_row.every(isCell)) {
        return null;
      }
      total_row = raw.total_row as ReportCell[];
    }
    return { type: 'table', title: raw.title as string, columns, rows, total_row, note };
  }
  return null;
}

/**
 * Narrow a stored report to the renderer's shape, or null.
 *
 * Checked at runtime rather than asserted, like `chatResultOf`: a report written
 * by an older handler, or a block the renderer does not know, renders as "no
 * report" rather than as a blank page with a header on it.
 */
export function reportSpecOf(raw: unknown): ReportSpec | null {
  if (!isRecord(raw)) return null;
  if (typeof raw.title !== 'string' || typeof raw.headline !== 'string') return null;
  if (typeof raw.period_start !== 'string' || typeof raw.period_end !== 'string') return null;
  if (typeof raw.period_label !== 'string') return null;
  if (!Array.isArray(raw.kpis) || !Array.isArray(raw.blocks)) return null;
  const kpis: ReportKpi[] = [];
  for (const k of raw.kpis) {
    const kpi = kpiOf(k);
    if (!kpi) return null;
    kpis.push(kpi);
  }
  const blocks: ReportBlock[] = [];
  for (const b of raw.blocks) {
    const block = blockOf(b);
    if (!block) return null;
    blocks.push(block);
  }
  return {
    title: raw.title,
    period_start: raw.period_start,
    period_end: raw.period_end,
    period_label: raw.period_label,
    headline: raw.headline,
    kpis,
    blocks,
  };
}
